torch/optim/swa.py

```diff
@@ -0,0 +1,371 @@
+            * param_groups - a dict containing all parameter groups
+        """
+        opt_state_dict = self._optimizer.state_dict()
+        swa_state = {(id(k) if isinstance(k, torch.Tensor) else k): v
+                     for k, v in self.state.items()}
+        opt_state = opt_state_dict["state"]
+        param_groups = opt_state_dict["param_groups"]
+        return {"opt_state": opt_state, "swa_state": swa_state,
+                "param_groups": param_groups}
+
+    def load_state_dict(self, state_dict):
+        r"""Loads the optimizer state.
+
+        Args:
+            state_dict (dict): SWA optimizer state. Should be an object returned
+                from a call to `state_dict`.
+        """
+        swa_state_dict = {"state": state_dict["swa_state"],
+                          "param_groups": state_dict["param_groups"]}
+        opt_state_dict = {"state": state_dict["opt_state"],
+                          "param_groups": state_dict["param_groups"]}
+        super(SWA, self).load_state_dict(swa_state_dict)
+        self._optimizer.load_state_dict(opt_state_dict)
+        self._opt_state = self._optimizer.state
+
+    def add_param_group(self, param_group):
+        r"""Add a param group to the :class:`Optimizer` s `param_groups`.
+
+        This can be useful when fine tuning a pre-trained network as frozen
+        layers can be made trainable and added to the :class:`Optimizer` as
+        training progresses.
+
+        Args:
+            param_group (dict): Specifies what Tensors should be optimized along
+            with group specific optimization options.
+        """
+        param_group['n_avg'] = 0
+        param_group['step_counter'] = 0
+        self._optimizer.add_param_group(param_group)
+
+    @staticmethod
+    def bn_update(loader, model, device=None):
+        r"""Updates BatchNorm running_mean, running_var buffers in the model.
+
+        It performs one pass over data in `loader` to estimate the activation
+        statistics for BatchNorm layers in the model.
+
+        Args:
+            loader (torch.utils.data.DataLoader): dataset loader to compute the
+                activation statistics on. Each data batch should be either a
+                tensor, or a list/tuple whose first element is a tensor
+                containing data.
+
+            model (torch.nn.Module): model for which we seek to update BatchNorm
+                statistics.
+
+            device (torch.device, optional): If set, data will be trasferred to
+                :attr:`device` before being passed into :attr:`model`.
+        """
+        if not _check_bn(model):
+            return
+        was_training = model.training
+        model.train()
+        momenta = {}
+        model.apply(_reset_bn)
+        model.apply(lambda module: _get_momenta(module, momenta))
+        n = 0
+        for input in loader:
+            if isinstance(input, (list, tuple)):
+                input = input[0]
+            b = input.size(0)
```

Comment:
This code doesn't work with dict-based datasets and I don't know how they could be supported here, so I'd suggest to raise a NotImplementedError that states this instead of failing with a not very helpful "KeyError: 0" message in line 316.

```suggestion
            elif isinstance(input, dict):
                raise NotImplementedError('Loaders that return dicts are not supported.')
            else:
                raise ValueError('Expected loader to contain tensors, lists or '
                                 'tuples of tensors, but got {}'.format(type(input)))
            b = input.size(0)
```